### PR TITLE
ICU-13417 Pass length from Locale::forLanguageTag() to ultag_parse().

### PR DESCRIPTION
**Pull request by roubert**
If not passed a length, ultag_parse() will call uprv_strlen(), requiring
the input buffer to be NUL terminated. This is unnecessary.

### CODE REVIEW

**Comment by roubert** [comment]
What do you think about doing a change along these lines? On one hand, the improvement is tiny. On the other hand, any improvement at all could be of value and this is something that would be very easy to do.

**Comment by markusicu** [comment]
- This kind of thing is ok in principle.
- It would want to be tagged @internal and use the hide-internal macro.
- Actually better to move it into an internal uloc_imp.h header (we probably have one like that) rather than declare+hide in the public one.
- While an improvement now, I wonder if it wouldn't be better to schedule time to rewrite the parsing in C++, with better internal API (StringPiece, ByteSink, ...), and turn the C functions into wrappers.
- More important would be to change internal call sites of Locale stuff to use the new API and stop dealing with fixed-size char[] buffers and string-not-terminated-warnings.

File: icu4c/source/common/unicode/uloc.h [hunk 1]
```diff
@@ -1142,6 +1142,16 @@ uloc_forLanguageTag(const char* langtag,
                     int32_t* parsedLength,
                     UErrorCode* err);
 
+#ifndef U_HIDE_SYSTEM_API
+U_STABLE int32_t U_EXPORT2
+uloc_forLanguageTagImpl(const char* langtag,
```
Comment (jefgen) [inline icu4c/source/common/unicode/uloc.h]: Could this go in `source\common\ulocimp.h` instead?

Comment (yumaoka) [reply icu4c/source/common/unicode/uloc.h]: ^ I agree.

**Comment by roubert** [comment]
OK, there seems to be some support for the idea to add such an internal helper function, so I've now made it into a proper change with the right header, naming and comments. Please review.

**Pull request merged**